git_guidelines.md

```diff
@@ -0,0 +1,269 @@
+|/
+*   55001c8d Merge pull request #something else
+```
+
+We create a new branch `feature_xyz` based on `master`, create two new commits
+`My commit 1` and `My commit 2` and submit a pull request into master. The log
+of the branch `feature_xyz` now looks like this:
+```
+$ git log feature_xyz --oneline --graph
+* 893fffa5 (HEAD -> feature_xyz) My commit 2
+* a2a22fb9 My commit 1
+*   efee9a2b (master) Merge pull request #something
+|\
+| * ead7f309 A commit on master
+|/
+*   55001c8d Merge pull request #something else
+```
+
+If now new commits are pushed to `master`, resulting in this log:
+```
+$ git log master --oneline --graph
+* 0d636cc9 (HEAD -> master) Hotfix for issue #something completely different
+*   efee9a2b Merge pull request #something
+|\
+| * ead7f309 A commit on master
+|/
+*   55001c8d Merge pull request #something else
+```
+then the branch `feature_xyz` is out of date with `master`, because it lacks the
+commit `0d636cc9`. We could now merge both branches (via the cli or GitHub's
+`Update Branch` button), but that will result in a messy history. Thus **don't**
+do it! If you do it, you'll have to remove the merge commits manually.
+
+Instead run: `git pull --rebase upstream master` in the `feature_xyz`
+branch. Git will pull the new commit `0d636cc9` from master into your branch
+`feature_xyz` and apply the two commits `My commit 1` and `My commit 2` on top
+of it:
+```
+$ git log feature_xyz --oneline --graph
+* 22a7a8c2 (HEAD -> feature_xyz) My commit 2
+* efe2ccdc My commit 1
+* 0d636cc9 (master) Hotfix for issue #something completely different
+*   efee9a2b Merge pull request #something
+|\
+| * ead7f309 A commit on master
+|/
+*   55001c8d Merge pull request #something else
+```
+Please note, that the hash of `My commit 1` and `My commit 2` changed! That
+happened because their parent changed. Therefore you have to force push your
+changes via `git push --force` next time you push your changes upstream.
+
+
+## Merging pull requests
+
+Most pull requests should be merged by creating a merge commit (the default on
+GitHub). Small pull requests (= only one can commit) can be rebased on top of
+master.
+
+
+## Branches and tags
+
+- The `master` branch is the current "main" development branch. It is protected
+  so that changes can be only included via reviewed pull requests. New releases
+  are made by tagging a specific commit on `master`.
+
+- Releases are tagged with a tag of the form `v$major.$minor`. The tag is not
+  changed when changes are backported.
+
+- For each release a branch of the form `$major.$minor` should be created to
+  store backported changes. It should be branched of from `master` at the commit
+  which was tagged with `v$major.$minor`.
+
+- All other branches are development branches for pull requests, experiments,
+  etc. They should be deleted once the pull request got merged or the branch is
+  no longer useful.
+
+- Exiv2 team members should create branches for pull requests in the main
```

Comment:
Should we always follow this pattern? At this moment, I am always using my own fork to collaborate. I think it is a protection mechanism against stupid human errors.

But I agree with you in which we could use branches in the main repository when collaboration is needed.

Reply:
Actually I am also using my own fork all the time, so I'll reword this a bit. Branches in the main repo probably only make sense if we are going to collaborate on something.